File: Python/ByRound/2172/2172_M_Maximum_Distance_To_Port.py
# Problem: CF 2172 M - Maximum Distance To Port
# https://codeforces.com/contest/2172/problem/M

"""
Algorithm: BFS to compute shortest distances from port (city 1) to all other cities.
Then, for each product type, find the maximum distance among all cities producing that product.

Time Complexity: O(n + m) - BFS traversal visits each node and edge once.
Space Complexity: O(n + m) - Storage for adjacency list and distance array.

The problem asks to find the longest shortest path from any city producing a specific product to the port city (city 1).
We use BFS from city 1 to compute all shortest distances, then for each product type,
we find the maximum distance among all nodes of that product type.
"""

n, m, k = map(int, input().split())
p = list(map(int, input().split()))  # Product types of each city
g = [[] for _ in range(n + 1)]       # Adjacency list representation of the graph
for _ in range(m):
    u, v = map(int, input().split())
    g[u].append(v)
    g[v].append(u)

d = [-1] * (n + 1)   # Distance array initialized to -1 (unvisited)
d[1] = 0             # Distance from port (city 1) to itself is 0
q = [1]              # Queue for BFS, starting with port city 1
h = 0                # Head pointer for the queue

# Perform BFS to compute distances from port city 1
while h < len(q):
    u = q[h]
    h += 1
    for v in g[u]:   # For each neighbor v of u
        if d[v] == -1:  # If v is unvisited
            d[v] = d[u] + 1  # Set distance to u's distance + 1
            q.append(v)      # Add v to queue for further exploration

r = [0] * (k + 1)    # Result array for maximum distance per product type
for i in range(1, n + 1):  # For each city
    product_type = p[i - 1]  # Product type of city i (0-indexed in p)
    r[product_type] = max(r[product_type], d[i])  # Update max distance for this product type

print(*r[1:])  # Output result for product types 1 to k


# https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/M/2172_M_Maximum_Distance_To_Port.cpp
// Problem: CF 2172 M - Maximum Distance To Port
// https://codeforces.com/contest/2172/problem/M

/*
 * Problem: M. Maximum Distance To Port
 * 
 * Purpose: Given a graph of cities connected by roads, each city produces one of k agricultural products.
 *          Find the maximum shortest distance from any city producing a specific product type to the port city (city 1).
 * 
 * Algorithms/Techniques:
 *   - BFS (Breadth-First Search) to compute shortest distances from port (city 1) to all other cities
 *   - For each product type, track the maximum distance among all cities producing that product
 * 
 * Time Complexity: O(n + m) where n is number of cities and m is number of roads.
 * Space Complexity: O(n + m) for adjacency list and auxiliary arrays.
 * 
 * Input:
 *   - n: number of cities
 *   - m: number of roads
 *   - k: number of product types
 *   - arr[i]: product type of city i
 *   - m lines of u v representing bidirectional roads
 * 
 * Output:
 *   - k integers representing max shortest distance for each product type to port
 */

#include <algorithm>
#include <iostream>
#include <queue>
#include <vector>

using namespace std;
const int N = 2e5 + 10;
int arr[N];
vector<int> g[N];
int n, m, k;
int ans[N];
int d[N];
queue<int> q;
bool st[N];

// BFS from city v to compute shortest distances to all reachable cities
void bfs(int v, int x) {
  q.push(v);
  st[v] = 1;
  d[v] = 0;
  while (q.size()) {
    int t = q.front();
    q.pop();
    for (int &i : g[t]) {
      if (st[i])
        continue;
      d[i] = d[t] + 1;
      q.push(i);
      st[i] = 1;
    }
  }
}

int main() {
  cin >> n >> m >> k;
  for (int i = 1; i <= n; i++)
    cin >> arr[i];
  for (int i = 1; i <= m; i++) {
    int u, v;
    cin >> u >> v;
    g[u].push_back(v);
    g[v].push_back(u);
  }
  bfs(1, 0); // Run BFS from the port (city 1)
  for (int i = 1; i <= n; i++)
    ans[arr[i]] = max(ans[arr[i]], d[i]); // Update max distance for product type
  for (int i = 1; i <= k; i++)
    cout << ans[i] << " ";
  return 0;
}


// https://github.com/VaHiX/CodeForces/